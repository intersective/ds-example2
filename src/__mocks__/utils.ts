export const Log = {
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
};
export const success = (x: any) => x;
export const error = jest.fn();
