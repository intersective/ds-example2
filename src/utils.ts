export import Log = require('@dazn/lambda-powertools-logger');
export const success = (data: any) => {
  return {
    success: true,
    data
  };
}

export const error = (data: any) => {
  return {
    success: false,
    data
  };
}

export const debug = Log.debug;
