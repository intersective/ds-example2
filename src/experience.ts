import { Log, success } from './utils';
import {
  readCache, writeCache,
  getEnrolmentCountQuery,
  getActiveUserCountQuery,
  getFeedbackLoopStartedCountQuery,
  getFeedbackLoopCompletedCountQuery,
  getReviewRatingAvgQuery,
  getTeamStatQuery,
  getAllExperiencesQuery,
} from './data-sources';

type Role = 'admin' | 'coordinator' | 'mentor' | 'participant';

const CACHE_KEY_PREFIX = 'experience-stats';

export const experience = async (id: number) => {
  const cacheKey = `${ CACHE_KEY_PREFIX }-${ id }`;
  Log.debug('reading cache for the experience', { cacheKey });
  const cachedValue = await readCache(cacheKey);
  Log.debug('get cached value for the experience', { cachedValue, now: Date.now() });
  // don't re-run the query if it is cached less than 1 minute ago
  if (cachedValue && (Date.now() - cachedValue.created < 1000 * 60)) {
    // need to return the data in the same format
    return success(cachedValue.data);
  }
  Log.debug('getting data from database', { id });
  const [
    { enrolledUserCount, registeredUserCount },
    activeUserCount,
    { feedbackLoopStarted, feedbackLoopCompleted },
    reviewRatingAvg,
    onTrackRatio
  ] = await Promise.all([
    getEnrolmentCount(id),
    getActiveUserCount(id),
    getFeedbackLoopCount(id),
    getReviewRatingAvg(id),
    getOnTrackRatio(id)
  ]);
  const result = {
    enrolledUserCount,
    registeredUserCount,
    activeUserCount,
    feedbackLoopStarted,
    feedbackLoopCompleted,
    reviewRatingAvg,
    onTrackRatio
  };
  Log.debug('Writting to cache', { cacheKey, result });
  writeCache(cacheKey, {
    created: Date.now(),
    data: result,
  });
  return success(result);
}

export const allExperiences = async () => {
  const exps: any[] = await getAllExperiencesQuery();
  let exp: any;
  const promises: Promise<any>[] = [];
  for (exp of exps) {
    // don't auto refresh completed experiences
    if (exp.status === 'completed') {
      continue;
    }
    const cacheKey = `${ CACHE_KEY_PREFIX }-${ exp.id }`;
    Log.debug('reading cache for each experience', { cacheKey });
    const cachedValue = await readCache(cacheKey);
    // don't refresh the data if it is cached less than 1 hour ago
    if (cachedValue && Date.now() - cachedValue.created < 1000 * 60 * 60) {
      continue;
    }
    Log.debug('refreshing data', { id: exp.id });
    promises.push(experience(exp.id));
  }
  await Promise.all(promises);
  Log.debug('get all the results');
}

async function getEnrolmentCount(experienceId: number) {
  Log.debug('getting enrolment count', { experienceId });
  const result = await getEnrolmentCountQuery(experienceId);
  const enrolledUserCount = {
    admin: 0,
    coordinator: 0,
    mentor: 0,
    participant: 0,
  };
  const registeredUserCount = {
    admin: 0,
    coordinator: 0,
    mentor: 0,
    participant: 0,
  };
  result.forEach((res: { role: Role; state: string; count: string; }) => {
    if (Object.prototype.hasOwnProperty.call(enrolledUserCount, res.role)) {
      enrolledUserCount[res.role] += +res.count;
      if (res.state === 'active') {
        registeredUserCount[res.role] += +res.count;
      }
    }
  });
  Log.debug('enrolment count data', { experienceId, enrolledUserCount, registeredUserCount });
  return { enrolledUserCount, registeredUserCount };
}

/**
 * Active user count
 *
 * Active user means a user that did any action within the last 7 days
 *
 */
async function getActiveUserCount(experienceId: number) {
  Log.debug('getting active user count', { experienceId });
  const result = await getActiveUserCountQuery(experienceId);
  const count = {
    admin: 0,
    coordinator: 0,
    mentor: 0,
    participant: 0,
  };
  result.forEach((res: { count: string; role: Role }) => {
    if (Object.prototype.hasOwnProperty.call(count, res.role)) {
      count[res.role] += +res.count;
    }
  });
  Log.debug('active user count data', { experienceId, count });
  return count;
}

/**
 * Feedback Loop
 *
 * Feedback loop started means a submission has been made.
 *   If it is a team submission, each team member count as one feedback loop
 *
 * Feedback loop completed means user has marked the todo item for seeing the feedback as done.
 *
 */
async function getFeedbackLoopCount(experienceId: number) {
  Log.debug('getting feedback loop count', { experienceId });
  const [ feedbackLoopStarted, feedbackLoopCompleted ] = await Promise.all([getFeedbackLoopStartedCountQuery(experienceId),
    getFeedbackLoopCompletedCountQuery(experienceId)
  ]);
  Log.debug('feedback loop count data', { experienceId, feedbackLoopStarted, feedbackLoopCompleted });
  return { feedbackLoopStarted, feedbackLoopCompleted };
}

function getReviewRatingAvg(experienceId: number) {
  Log.debug('getting review rating', { experienceId });
  return getReviewRatingAvgQuery(experienceId);
}

/**
 * On track ratio
 *
 * User is on track means (no. of on track records)/(no. of off track records) >= 0.8
 *
 * On track ratio = (no. of user on track)/(no. of user on track) + (no. of user off track)
 */
async function getOnTrackRatio(experienceId: number) {
  Log.debug('getting on track ratio', { experienceId });
  const teamStats = await getTeamStatQuery(experienceId);
  if (!teamStats.length) {
    return -1;
  }
  interface UserStat {
    id: number;
    on: number;
    off: number;
  }
  const userStats: UserStat[] = [];
  teamStats.forEach((teamStat: { user_id: number; stat_value: string; }) => {
    const userIndex = userStats.findIndex((u: UserStat) => u.id === teamStat.user_id);
    if (userIndex >= 0) {
      if (teamStat.stat_value === '1') {
        userStats[userIndex].on += 1;
      } else {
        userStats[userIndex].off += 1;
      }
    } else {
      userStats.push({
        id: teamStat.user_id,
        on: teamStat.stat_value === '1' ? 1 : 0,
        off: teamStat.stat_value === '0' ? 1 : 0,
      });
    }
  });
  const onTrackUserCount = userStats.filter((u: UserStat) => u.on/(u.on + u.off) >= 0.8).length;
  Log.debug('on track ratio data', { experienceId, onTrackUserCount });
  return +(onTrackUserCount/userStats.length).toFixed(2);
}
