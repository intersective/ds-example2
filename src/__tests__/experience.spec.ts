import { experience, allExperiences } from '../experience';
import { success } from '../utils';
import {
  readCache,
  getEnrolmentCountQuery,
  getActiveUserCountQuery,
  getFeedbackLoopStartedCountQuery,
  getFeedbackLoopCompletedCountQuery,
  getReviewRatingAvgQuery,
  getTeamStatQuery,
  getAllExperiencesQuery,
} from '../data-sources';
jest.mock('../utils');
jest.mock('../data-sources');

beforeEach(() => {
  (readCache as jest.Mock).mockReturnValue(null);
  (getEnrolmentCountQuery as jest.Mock).mockReturnValue([
    {
      role: 'participant',
      state: 'active',
      count: 3
    },
    {
      role: 'participant',
      state: 'unverified',
      count: 5
    },
    {
      role: 'mentor',
      state: 'active',
      count: 2
    },
    {
      role: 'mentor',
      state: 'unverified',
      count: 6
    },
    {
      role: 'coordinator',
      state: 'active',
      count: 4
    },
    {
      role: 'coordinator',
      state: 'unverified',
      count: 7
    },
    {
      role: 'admin',
      state: 'active',
      count: 1
    },
    {
      role: 'admin',
      state: 'unverified',
      count: 3
    }
  ]);
  (getActiveUserCountQuery as jest.Mock).mockReturnValue([
    {
      role: 'participant',
      count: 20
    },
    {
      role: 'mentor',
      count: 5
    },
    {
      role: 'coordinator',
      count: 4
    },
    {
      role: 'admin',
      count: 2
    }
  ]);
  (getFeedbackLoopStartedCountQuery as jest.Mock).mockReturnValue(15);
  (getFeedbackLoopCompletedCountQuery as jest.Mock).mockReturnValue(4);
  (getReviewRatingAvgQuery as jest.Mock).mockReturnValue(0.4);
  (getTeamStatQuery as jest.Mock).mockReturnValue([
    {
      user_id: 1,
      stat_value: '1'
    },
    {
      user_id: 1,
      stat_value: '1'
    },
    {
      user_id: 1,
      stat_value: '1'
    },
    {
      user_id: 1,
      stat_value: '1'
    },
    {
      user_id: 1,
      stat_value: '1'
    },
    {
      user_id: 1,
      stat_value: '0'
    },
    {
      user_id: 2,
      stat_value: '1'
    },
    {
      user_id: 2,
      stat_value: '1'
    },
    {
      user_id: 2,
      stat_value: '0'
    },
    {
      user_id: 3,
      stat_value: '1'
    },
    {
      user_id: 3,
      stat_value: '0'
    },
    {
      user_id: 4,
      stat_value: '0'
    }
  ]);
  (getAllExperiencesQuery as jest.Mock).mockReturnValue([
    {
      id: 1,
      status: 'live'
    },
    {
      id: 2,
      status: 'completed'
    }
  ]);
});

afterEach(async () => {
  const res = await experience(1);
  expect(res).toMatchSnapshot();
  allExperiences();
});

it('1. get cached value', () => {
  (readCache as jest.Mock).mockReturnValueOnce({
    created: Date.now() - 1,
    data: 'cached data'
  });
});

it('2. read data from database', () => {

});

it('3. no team stats', () => {
  (getTeamStatQuery as jest.Mock).mockReturnValueOnce([]);
  (readCache as jest.Mock).mockReturnValue({
    created: Date.now() - 1000 * 61,
    data: 'cached data'
  });
});
