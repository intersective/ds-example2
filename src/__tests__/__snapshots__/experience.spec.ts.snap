// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. get cached value 1`] = `"cached data"`;

exports[`2. read data from database 1`] = `
Object {
  "activeUserCount": Object {
    "admin": 2,
    "coordinator": 4,
    "mentor": 5,
    "participant": 20,
  },
  "enrolledUserCount": Object {
    "admin": 4,
    "coordinator": 11,
    "mentor": 8,
    "participant": 8,
  },
  "feedbackLoopCompleted": 4,
  "feedbackLoopStarted": 15,
  "onTrackRatio": 0.25,
  "registeredUserCount": Object {
    "admin": 1,
    "coordinator": 4,
    "mentor": 2,
    "participant": 3,
  },
  "reviewRatingAvg": 0.4,
}
`;

exports[`3. no team stats 1`] = `
Object {
  "activeUserCount": Object {
    "admin": 2,
    "coordinator": 4,
    "mentor": 5,
    "participant": 20,
  },
  "enrolledUserCount": Object {
    "admin": 4,
    "coordinator": 11,
    "mentor": 8,
    "participant": 8,
  },
  "feedbackLoopCompleted": 4,
  "feedbackLoopStarted": 15,
  "onTrackRatio": -1,
  "registeredUserCount": Object {
    "admin": 1,
    "coordinator": 4,
    "mentor": 2,
    "participant": 3,
  },
  "reviewRatingAvg": 0.4,
}
`;
